### PR TITLE
melange: the compiler does not support -bin-annot-occurrences

### PR DESCRIPTION
**Pull request by voodoos**
The Melange compiler does not accept the `-bin-annot-occurrences` flag.

The test relying on OCaml's version is therefore incorrect when building with melange compiler.
This PR adds a special case to prevent use of this flag when building with melange.

### CODE REVIEW

**Comment by emillon** [comment]
cc @anmonteiro - do we want this, or fix this on the melange side?
(@gridbugs we might need to backport this to 3.16 if we go with the patch route)

**Comment by anmonteiro** [comment]
We’ll have to fix it in both sides I suppose. Where can I read more about this flag?

File: src/dune_rules/module_compilation.ml [hunk 1]
```diff
@@ -183,23 +183,20 @@ let build_cm
...
-         in
-         fn :: other_targets, As annots)
-       else other_targets, Command.Args.empty
+     match Compilation_context.bin_annot cctx, cm_kind with
+     | false, _ | true, Ocaml Cmx -> other_targets, Command.Args.empty
+     | true, Ocaml (Cmi | Cmo) ->
+       let fn = Option.value_exn (Obj_dir.Module.cmt_file obj_dir m ~cm_kind ~ml_kind) in
+       let annots =
+         if Version.supports_bin_annot_occurrences ocaml.version
+         then [ "-bin-annot"; "-bin-annot-occurrences" ]
```
Comment (rgrinberg) [inline src/dune_rules/module_compilation.ml:192]: spurious change? the previous version did not duplicate the flag name here

File: src/dune_rules/module_compilation.ml [hunk 2]
```diff
@@ -183,23 +183,20 @@ let build_cm
...
-     | Ocaml (Cmi | Cmo) | Melange (Cmi | Cmj) ->
-       if Compilation_context.bin_annot cctx
-       then (
-         let fn =
-           Option.value_exn (Obj_dir.Module.cmt_file obj_dir m ~cm_kind ~ml_kind)
-         in
-         let annots =
-           [ "-bin-annot" ]
-           @
-           if Version.supports_bin_annot_occurrences ocaml.version
```
Comment (rgrinberg) [inline src/dune_rules/module_compilation.ml:197]: How about just maintaining the old code but introducing:

```
let supports_bin_annot_occurences = Version.supports_bin_annot_occurences ocaml.version && not cm_kind <> Melange in
```

**Review by rgrinberg — CHANGES_REQUESTED** [review]

**Comment by anmonteiro** [comment]
actually, since melange supports 5.2, this is just a matter of supporting the flag. 

I bet it'd be nice to have occurrences support in Melange projects, too, so let me release a Melange version with support for this flag.

**Comment by anmonteiro** [comment]
released melange 4.0.1-52 with this change: https://github.com/ocaml/opam-repository/pull/26043

**Comment by voodoos** [comment]
Thanks @anmonteiro

I guess that means this PR is no more useful ?

**Comment by anmonteiro** [comment]
I’d vote to close it

**Pull request closed without merging**